api/v1/storage_autoscaler.go

```diff
@@ -0,0 +1,128 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1
+
+import (
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// StorageAutoScalerSpec defines the desired state of StorageAutoScaler
+type StorageAutoScalerSpec struct {
+	// StorageCluster is the name of the storage cluster for which the storage scaling is to be done.
+	// +kubebuilder:validation:Required
+	StorageCluster corev1.LocalObjectReference `json:"storageCluster,omitempty"`
+	// DeviceClass is the name of the device class for which the storage scaling is to be done.
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:default="ssd"
+	DeviceClass string `json:"deviceClass,omitempty"`
+	// StorageCapacityLimit is the total aggregate capacity limit for the storage scaling for the specific deviceClass and storagecluster.
+	// +kubebuilder:validation:Required
+	StorageCapacityLimit resource.Quantity `json:"storageCapacityLimit,omitempty"`
+	// MaxOsdSize is the maximum size that Osd disk can be expanded to.
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:default="8Ti"
+	MaxOsdSize resource.Quantity `json:"maxOsdSize,omitempty"`
+	// StorageScalingThresholdPercent is the threshold percentage of the storage capacity that triggers the auto-scaling of the OSDs.
+	// Should be less than the OsdNearFullThresholdPercentage.
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:default=70
+	StorageScalingThresholdPercent int `json:"storageScalingThresholdPercent,omitempty"`
+	// TimeoutSeconds is the time in seconds after which the storage auto-scaler will alert the user that the scaling operation has been failed.
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:default=1800
+	TimeoutSeconds int `json:"timeoutSeconds,omitempty"`
+}
+
+type StorageAutoScalerPhase string
+
+const (
+	StorageAutoScalerPhaseNotStarted StorageAutoScalerPhase = "NotStarted"
+	StorageAutoScalerPhaseInProgress StorageAutoScalerPhase = "InProgress"
+	StorageAutoScalerPhaseSucceeded  StorageAutoScalerPhase = "Succeeded"
+	StorageAutoScalerPhaseFailed     StorageAutoScalerPhase = "Failed"
+)
+
+// TimestampedError is the error message with the time stamp.
+type TimestampedError struct {
+	// Message is the error message in case the storage scaling operation has failed.
+	Message string `json:"message,omitempty"`
+	// Timestamp is the time stamp when the error occurred.
+	Timestamp metav1.Time `json:"timestamp,omitempty"`
+}
+
+// StorageAutoScalerStatus defines the observed state of StorageAutoScaler
+type StorageAutoScalerStatus struct {
+	// Phase describes the Phase of StorageAutoScaler
+	// +kubebuilder:validation:Enum="";'NotStarted;InProgress;Failed;Succeeded
+	Phase StorageAutoScalerPhase `json:"phase,omitempty"`
```

Comment:
I think a previous note I added as a reply got missed. It'd be best to not enum the phase. This can only introduce errors creating the CRD, and I don't know of any resources that enum status items like this.

For comparison the StorageCluster status phase isn't enum-ed

https://github.com/red-hat-storage/ocs-operator/blob/be77c56795c42217cbd4c55b1d3b41039cf500be/api/v1/storagecluster_types.go#L545-L548

Reply:
why cant we use enum with `+ nullable`?

Reply:
I'm not sure if we can use enum with nullable, since nullable implies a type that is a pointer. It might be valid or might not. But in general, I don't think enums in status items are very helpful. In my experience, they are unlikely to help and likely to cause a problem, so don't bother. They won't help users, since users don't set the status themselves. Also, they can result in the operator being unable to set otherwise valid statuses, resulting in unexpected runtime errors that we need to bug fix.

Reply:
Actually I removed +nullable, and used a default value,

The main reason here is to use enum is that coming from alert team, as alert are based on the phase they had to make sure operator is setting up correctly

Reply:
Updated